### DISCUSSION TITLE
How how can I use react-hook-form in tree data type

### DISCUSSION THREAD

**Discussion by Cyberhan123** [react-hook-form/react-hook-form#4103] [Q&A] [answered]
The data type like this
```
const son = {
    sonProps1: '',
    sonProps2: '',
};


const father= {
    fatheProps1: '',
    fatheProps2: '',
    children: [],
};

const grandfather= {
    grandfatheProps1: '',
    children: [],
};
```
son can push into father.children array aslo father can push into grandfather.children

user can delete instert each tree leve,how can i work with react-hook-form? my company use an custom react ui components，
i use Controller wrap them，and they can't get ref，so it so hard to code ... please help me,thx!!!

**Comment by bluebill1049** [comment, marked as answer]
are you after https://react-hook-form.com/api#useFieldArray ?

**Reply by Cyberhan123** [reply]
I see,but the doc notice me :useFieldArray is a custom hook for working with uncontrolled Field Arrays... emmm, it can work with useController? Can u give a example?

**Reply by bluebill1049** [reply]
you build a component with `useController` and then just slot it in. same thing right?

**Reply by Cyberhan123** [reply]
> you build a component with `useController` and then just slot it in. same thing right?

I have an confusion ，useArray can't use in each node，it will case react warning: Rendered more hooks than during the previous render.

The reason for triggering is that recursion cannot be declared useArray Hook

**Reply by phinguyen-zoi** [reply]
Hi @Cyberhan123 ,
Could you share with us the solution you end up with?

**Comment by Cyberhan123** [comment]
My question is quite special. I have written it completely. @bluebill1049 thx , I'll look at that later
